crates/cheatcodes/spec/src/vm.rs

```diff
@@ -260,6 +260,26 @@ interface Vm {
         uint64 depth;
     }
 
+    /// The result of the `stopDebugTraceRecording` call
+    struct DebugStep {
+        /// The stack before executing the step of the run.
+        /// stack\[0\] represents the top of the stack.
+        /// and only stack data relevant to the opcode execution is contained.
+        uint256[] stack;
+        /// The memory data before executing the step of the run.
+        /// only data relevant to the opcode execution is contained.
+        /// e.g. for MLOAD, it will have 32 bytes of the memory data here.
+        uint8[] memoryData;
```

Comment:
let's change this to `bytes`

is there a usecase for those opcodes which only requires memory input slice and nothing else? there's also memory we write to, max memory expansion size, returndata which are all not included here, though I guess some of those might cause too high memory usage

adding any of those later would be a breaking change so unsure what's best way to handle this

Reply:
changed to bytes in https://github.com/foundry-rs/foundry/pull/8571/commits/562691fe90e6b71ffab84590dc7d40920d4e3392

> is there a usecase for those opcodes which only requires memory input slice and nothing else? 

There are two use cases needed in validating ERC4337 process:
1. To get the function selector bytes for the CALL/STATICCALL opcodes. This is to whitelist some functions that can be called on EntryPoint contract, which most functions are supposedly not able to be called.
2. To get the input used in keccak256. The reason is that the "associated storage" like mapping are allowed, but the storage slots are some result of the sha. We need to whitelist those storage slot accesses while ban any other ones. Since we cannot reverse the sha, we can only get those from the memory input.

I change the variable name to `memoryInput`, hopefully making it a bit less conflicting with potential future use cases of this DebugStep struct. e.g., you can add `returnData` in the future as new field if wanted.

> memory we write to

The location can be get via the stack data, and the memory result will mostly be what we are writting to anyway (input)

> max memory expansion size

I am guessing you mean free pointer? if needed I guess we can add a new field for this as well in the future as extenstion?

Reply:
yeah, I guess we can just add all of those later in backwards-compatible manner, so let's firstly get the initial implementation working